### PR TITLE
[Paddle Inference] General optimization for no_varlen skiplayernorm 

### PR DESCRIPTION
**Pull request by Wangzheee**
### PR types

Others
### PR changes

Others
### Describe

General optimization for no_varlen embedding layernorm:

- Use VarSeqlen kernel
- Support fp16, fp32
- Improve performance

### CODE REVIEW

File: paddle/fluid/framework/ir/trt_skip_layernorm_fuse_pass.cc [hunk 1]
```diff
@@ -169,8 +169,21 @@ void TrtSkipLayerNormFusePass::ApplyImpl(ir::Graph *graph) const {
...
+          static_cast<int32_t>(elementwise_out->Var()->GetShape().size());
+      if ((begin_norm_axis != -1) && (begin_norm_axis != input_rank - 1)) {
+        LOG(WARNING) << "skip_layernorm pass only support "
+                        "layer_norm'begin_norm_axis == input_rank - 1.";
+        return;
+      }
+      new_desc.SetAttr("begin_norm_axis", begin_norm_axis);
+    }
+    int32_t ld = layer_norm_scale->Var()->GetShape()[0];
+    new_desc.SetAttr("ld", ld);
```
Comment (qingqing01) [inline paddle/fluid/framework/ir/trt_skip_layernorm_fuse_pass.cc]: ld is leading dimension ?  this name is hard to understand here

File: paddle/fluid/inference/tensorrt/convert/skip_layernorm.cc [hunk 2]
```diff
@@ -26,183 +25,132 @@ class SkipLayerNormOpConverter : public OpConverter {
...
+           bias_weight.values,
+           GetPluginFieldType(bias_weight.type),
+           static_cast<int32_t>(bias_weight.count)},
+          {"gamma",
+           scale_weight.values,
+           GetPluginFieldType(scale_weight.type),
+           static_cast<int32_t>(scale_weight.count)}};
+      nvinfer1::PluginFieldCollection* pluginPtr =
+          static_cast<nvinfer1::PluginFieldCollection*>(
+              malloc(sizeof(*pluginPtr) +
```
Comment (qingqing01) [inline paddle/fluid/inference/tensorrt/convert/skip_layernorm.cc]: sizeof(*pluginPtr)  ->  sizeof(nvinfer1::PluginFieldCollection) ?

File: paddle/fluid/inference/tensorrt/convert/skip_layernorm.cc [hunk 3]
```diff
@@ -26,183 +25,132 @@ class SkipLayerNormOpConverter : public OpConverter {
...
+      PADDLE_ENFORCE_NE(
+          creator,
+          nullptr,
+          platform::errors::InvalidArgument(
+              "fail to get creator of CustomSkipLayerNormPluginDynamic"));
+      int32_t type = static_cast<int32_t>((engine_->WithFp16() == 1)
+                                              ? nvinfer1::DataType::kHALF
+                                              : nvinfer1::DataType::kFLOAT);
+      if (enable_int8) {
+        type = static_cast<int32_t>(nvinfer1::DataType::kHALF);
```
Comment (qingqing01) [inline paddle/fluid/inference/tensorrt/convert/skip_layernorm.cc:114]: add comments why set type kHALF when enable INT8

File: paddle/fluid/inference/tensorrt/convert/skip_layernorm.cc [hunk 4]
```diff
@@ -26,183 +25,132 @@ class SkipLayerNormOpConverter : public OpConverter {
...
+          {"gamma",
+           scale_weight.values,
+           GetPluginFieldType(scale_weight.type),
+           static_cast<int32_t>(scale_weight.count)},
+      };
+      nvinfer1::PluginFieldCollection* pluginPtr =
+          static_cast<nvinfer1::PluginFieldCollection*>(
+              malloc(sizeof(*pluginPtr) +
+                     fields.size() *
+                         sizeof(nvinfer1::PluginField)));  // remember to free
```
Comment (qingqing01) [inline paddle/fluid/inference/tensorrt/convert/skip_layernorm.cc:140]: remeber to free ? where to release pluginPtr?

File: python/paddle/fluid/tests/unittests/ir/inference/test_trt_skip_layernorm_fuse_pass.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+class SkipLayernormFusePassTest0(InferencePassTest):
+    def setUp(self):
+        with fluid.program_guard(self.main_program, self.startup_program):
+            data1 = fluid.data(
+                name="data1", shape=[-1, 3, 128, 128], dtype="float32"
+            )
+            data2 = fluid.data(
+                name="data2", shape=[-1, 3, 128, 128], dtype="float32"
+            )
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/ir/inference/test_trt_skip_layernorm_fuse_pass.py:36]: fluid.data -> paddle.static.data

File: python/paddle/fluid/tests/unittests/ir/inference/test_trt_skip_layernorm_fuse_pass.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+            eltwise_out = self.append_eltwise(data1, data2)
+            out = fluid.layers.layer_norm(input=eltwise_out, begin_norm_axis=3)
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/ir/inference/test_trt_skip_layernorm_fuse_pass.py]: use paddle.nn.functional.layer_norm, but note `normalized_shape` is used to set norm axis

File: python/paddle/fluid/tests/unittests/ir/inference/test_trt_skip_layernorm_fuse_pass.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+                {'data1': [1, 1, 1, 128], 'data2': [1, 1, 1, 128]},
+                {'data1': [1, 3, 128, 128], 'data2': [1, 3, 128, 128]},
+                {'data1': [1, 3, 128, 128], 'data2': [1, 3, 128, 128]},
+                False,
+            )
+        )
+        self.fetch_list = [out]
+
+    def append_eltwise(self, data1, data2):
+        return fluid.layers.elementwise_add(x=data1, y=data2)
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/ir/inference/test_trt_skip_layernorm_fuse_pass.py]: fluid.layers.elementwise_add -> paddle.add

File: python/paddle/fluid/tests/unittests/ir/inference/inference_pass_test.py [hunk 8]
```diff
@@ -36,7 +36,7 @@ def __init__(self, methodName='runTest'):
         self.enable_mkldnn = False
         self.enable_mkldnn_bfloat16 = False
         self.enable_trt = False
-        self.enable_tensorrt_varseqlen = True
+        self.enable_tensorrt_varseqlen = False
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/ir/inference/inference_pass_test.py:39]: Why change this default argument?

File: paddle/fluid/framework/ir/trt_skip_layernorm_fuse_pass.cc (hunk 1 above)
Comment (Wangzheee) [reply paddle/fluid/framework/ir/trt_skip_layernorm_fuse_pass.cc]: ld是对应kernel的矩阵乘的参数ld，如果对应到模型上，可能用hidden_size更合适？

File: paddle/fluid/inference/tensorrt/convert/skip_layernorm.cc (hunk 3 above)
Comment (Wangzheee) [reply paddle/fluid/inference/tensorrt/convert/skip_layernorm.cc:114]: 其实有int8的kernel，slim同学觉得int8的kernel精度太差，在开启int8的时候调用fp16的kernel

File: python/paddle/fluid/tests/unittests/ir/inference/inference_pass_test.py (hunk 8 above)
Comment (Wangzheee) [reply python/paddle/fluid/tests/unittests/ir/inference/inference_pass_test.py:39]: 变长默认不开启，修复单测的bug

File: paddle/fluid/inference/tensorrt/convert/skip_layernorm.cc (hunk 2 above)
Comment (Wangzheee) [reply paddle/fluid/inference/tensorrt/convert/skip_layernorm.cc]: done

File: paddle/fluid/inference/tensorrt/convert/skip_layernorm.cc (hunk 4 above)
Comment (Wangzheee) [reply paddle/fluid/inference/tensorrt/convert/skip_layernorm.cc:140]: done

File: paddle/fluid/inference/tensorrt/convert/skip_layernorm.cc [hunk 9]
```diff
@@ -26,183 +25,132 @@ class SkipLayerNormOpConverter : public OpConverter {
...
-        auto plugin_layer = engine_->network()->addPluginV2(
-            inputs.data(), inputs.size(), *pluginObj);
-
-        PADDLE_ENFORCE_NE(
-            plugin_layer,
-            nullptr,
-            platform::errors::InvalidArgument(
-                "fail to add CustomSkipLayerNormPluginDynamic layer"));
-        layer = plugin_layer;
+    if (flag_varseqlen && engine_->with_interleaved()) {
```
Comment (fengxiaoshuai) [inline paddle/fluid/inference/tensorrt/convert/skip_layernorm.cc:59]: skip_layernorm应该算是一个常用结构，为什么这里会牵扯到ernie的配置posid和maskid？而且layer里也没用这两个参数

Comment (Wangzheee) [reply paddle/fluid/inference/tensorrt/convert/skip_layernorm.cc:59]: 这个是with_interleaved 排布的kernel，有个专门的分支，这个分支必须是变长+int8

File: python/paddle/fluid/tests/unittests/ir/inference/test_trt_skip_layernorm_fuse_pass.py (hunk 5 above)
Comment (Wangzheee) [reply python/paddle/fluid/tests/unittests/ir/inference/test_trt_skip_layernorm_fuse_pass.py:36]: done

File: python/paddle/fluid/tests/unittests/ir/inference/test_trt_skip_layernorm_fuse_pass.py (hunk 6 above)
Comment (Wangzheee) [reply python/paddle/fluid/tests/unittests/ir/inference/test_trt_skip_layernorm_fuse_pass.py]: done

**Review by zhangjun — DISMISSED** [review]
LGTM

**Review by qingqing01 — DISMISSED** [review]
LGTM

**Review by zhangjun — DISMISSED** [review]
LGTM

**Review by XieYunshen — DISMISSED** [review]
LGTM

**Review by qingqing01 — DISMISSED** [review]

**Review by XieYunshen — APPROVED** [review]
